### PR TITLE
Add enabled as a source config

### LINKED ISSUE #3662: [CT-201] Reconcile configs + properties for sources
Follow-up to #2401, #3616

Let's convert some `source` properties into **configs** that can be set inside `config:` blocks and within `dbt_project.yml`:

```yml
# models/src_whatever.yml
version: 2
sources:
  - name: my_source
    description: ... # not a config
    config:
      enabled: ...
      quoting: {dict}
      freshness: {dict}
      loader: ...
      loaded_at_field: ...
      database: ... # or 'project' in dbt-bigquery
      schema: ... # or 'dataset' in dbt-bigquery
      identifier: ... # this is like an alias for alias
      meta: {dict}
      tags: ...
    tables:
      - name: my_src_table
        config: # all the same stuff as above. these take precedence for specific tables
        description: ... # not a config
        tests: ... # not a config
        columns: ... # not a config
```

```yml
# dbt_project.yml
sources:
  project_name:
    subdirectory:
      +database: raw
      +loader: fivetran
      another_subdirectory:
        +enabled: false
```

For backwards compatibility, we should still support setting these as top-level properties:
```yml
sources:
  - name: my_source
    loaded_at_field: updated_at
```

But raise an error if the same config is set in both places (even if it's with the same value):
```yml
sources:
  - name: my_source
    loaded_at_field: updated_at
    config:
      loaded_at_field: updated_at
```

### Notes

- We're thinking that descriptions shouldn't be configs. They're rendered with a different context (docs), and they don't make sense to set hierarchically. The same goes for `tests` and `columns`—this would be quite tricky to figure out, since `tests` actually generate _new_ nodes, rather than adding properties to the existing node.

- As far as the manifest / `graph.sources` context, I'm open to suggestions. For backwards compatibility, we'd want to store things like `loaded_at_field` in both `node.config` _and_ node-level keys. But I think there's a valid argument for removing this as a top-level key and only storing it in `node.config`, so long as we communicate clearly that such a move is taking place.

- For better `state:modified` comparisons, we'd want to store the un-rendered version of these configs in `node.unrendered_config`, regardless of whether they're set in `dbt_project.yml` or `models/src_whatever.yml`. Original issue for this is https://github.com/dbt-labs/dbt/issues/2744.

### PR DESCRIPTION
**Pull request by emmyoop**
resolves #3662

### Description

This only adds enabled as a config.  Tests have been slimmed down to show what is expected to work.  Since you can't set enabled as a property currently, we don't need to deal with overrides.  `enabled` as a config is also highly valuable to the users of dbt.

The rest of the configs will be added in #5003.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have added information about my change to be included in the [CHANGELOG](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry).

### CODE REVIEW

**Review by jtcohen6 — COMMENTED** [review]
Whoops, I left my comments in https://github.com/dbt-labs/dbt-core/pull/5003 instead of here. Some of the context there may still be helpful/interesting to you, specifically the piece about `unrendered_config`.

In any case, I think you're on the right track here! Looks good in local testing. I'd be very happy about shipping this as a standalone piece of net-new functionality in v1.1.

File: core/dbt/parser/sources.py [hunk 1]
```diff
@@ -261,19 +259,27 @@ def parse_source_test(
...
             generator = ContextConfigGenerator(self.root_project)
         else:
             generator = UnrenderedConfigGenerator(self.root_project)
 
+        # configs with precendence set
+        precedence_configs = dict()
+        # first apply source configs
+        precedence_configs.update(target.source.config)
+        # then overrite anything that is defined on source tables
+        precedence_configs.update(target.table.config)
```
[the remark this answers is not in the document]
Comment (nathaniel-may) [reply core/dbt/parser/sources.py:274]: yeah this is the part that I really wasn't sure was a good idea. Is there a `merge_config_dict` function of sorts that will handle some of this?

**Comment by nathaniel-may** [comment]
blocked on #5015

**Comment by nathaniel-may** [comment]
looking into the one remaining failure.

**Review by emmyoop — COMMENTED** [review]
This looks good to me as a first step to the end goal.  I'm not going to approve it since I wrote some of it too!  I think it would be great if @gshank could take a look.

File: core/dbt/contracts/graph/unparsed.py [hunk 2]
```diff
@@ -242,6 +242,7 @@ class Quoting(dbtClassMixin, Mergeable):
 
 @dataclass
 class UnparsedSourceTableDefinition(HasColumnTests, HasTests):
+    config: Dict[str, Any] = field(default_factory=dict)
```
Comment (gshank) [inline core/dbt/contracts/graph/unparsed.py:245]: There is a HasConfig mixin class which some of the other Unparsed nodes use. If it's possible to use that (sometimes you run into precedence issues with fields without defaults) it probably should for consistency.

File: core/dbt/parser/sources.py [hunk 3]
```diff
@@ -261,19 +259,29 @@ def parse_source_test(
...
         else:
             generator = UnrenderedConfigGenerator(self.root_project)
 
+        # configs with precendence set
+        precedence_configs = dict()
+        # first apply source configs
+        precedence_configs.update(target.source.config)
+        # then overrite anything that is defined on source tables
+        # this is not quite complex enough for configs that can be set as top-level node keys, but
+        # it works while source configs can only include `enabled`.
```
[the remark this answers is not in the document]
Comment (gshank) [reply core/dbt/parser/sources.py:275]: In the other nodes configs are merged by the code in the BaseConfig object, in the 'update_from' method, which handles merging things like tags and meta, etc. (MergeBehavior, etc). Which is called by 'calculate_node_config'.  I imagine you'll need to use some of that code to handle merging the Source configs with the SourceTable configs?

**Review by gshank — APPROVED** [review]
Looks good. I left a few comments. The merging of configs will have to be handled eventually but that can happen in the next phase.

File: core/dbt/contracts/graph/unparsed.py (hunk 2 above)
Comment (emmyoop) [reply core/dbt/contracts/graph/unparsed.py:245]: We ran into issues with defaults when we tried to use it.

Comment (gshank) [reply core/dbt/contracts/graph/unparsed.py:245]: I'm not surprised. Getting the default and non-default configs lined up is a headache.

**Comment by nathaniel-may** [comment]
Also not approving since I wrote some of this as well. Thanks for the review, @gshank!

**Pull request merged**